substrate/client/consensus/beefy/src/lib.rs

```diff
@@ -525,10 +554,13 @@ pub async fn start_beefy_gadget<B, BE, C, N, P, R, S, AuthorityId>(
 
 	let metrics = register_metrics(prometheus_registry.clone());
 
+	let mut block_import_justif = links.from_block_import_justif_stream.subscribe(100_000).fuse();
+
 	// Subscribe to finality notifications and justifications before waiting for runtime pallet and
 	// reuse the streams, so we don't miss notifications while waiting for pallet to be available.
-	let mut finality_notifications = client.finality_notification_stream().fuse();
-	let mut block_import_justif = links.from_block_import_justif_stream.subscribe(100_000).fuse();
+	let finality_notifications = client.finality_notification_stream();
+	let (mut transformer, mut finality_notifications) =
+		finality_notification_transformer_future(finality_notifications);
```

Comment:
Would it work if we just did something like:
```
let mut finality_notifications = client
		.finality_notification_stream()
		.map(|finality_notification| UnpinnedFinalityNotification::from(finality_notification))
		.fuse();
```
?

And in the function definitions we could use:
```
		finality_notifications: &mut Fuse<
			impl Stream<Item = UnpinnedFinalityNotification<B>> + Unpin,
		>
```

Reply:
> Would it work if we just did something like:
> 
> ```
> let mut finality_notifications = client
> 		.finality_notification_stream()
> 		.map(UnpinnedFinalityNotification::from)
> 		.fuse();
> ```
> 
> ?

This would map the stream, but not solve our problem. The mapping takes place when the next item is retrieved from the stream. Our problem is that `next()` is not called while we are waiting for the header ancestry to become available. We need an extra future that runs concurrently for this mapping to happen in the background.

Reply:
Oh, I see, you're right. This makes sense. But, at a first glance, personally I think it would be better if instead we had a wrapper over the finality notifications stream with some internal state. Something like:

```
FinalityStream {
    inner: original finality stream,
    mode: Limited | Full
    queue: queued notifications while in Limited mode
}
```

We could have it in Limited mode while waiting for the header ancestry to become available, and then switching it to Full. In Limited mode we can just take the notifications, transform them to unpinned and push them to a queue. We still need to pump it while waiting for `async_initialize()` to finish, just as you do with the transformer here.

Also maybe we could drop all the finality notifications, except for the ones for mandatory headers while in Limited mode ?

Anyway, not a blocker on my side. If the current approach works, I'm ok with merging it, and then I can experiment with this wrapper.

Reply:
I mean we could do that, but I am not sure if there is any benefit. This transformer is already super simple and gets the job done. I think we can merge the current approach and then iterate on it, for now my priority is to not pin unnecessary blocks. Right now I am warp syncing some nodes to make sure nothing unexpected happens.

> Also maybe we could drop all the finality notifications, except for the ones for mandatory headers while in 
> Limited mode ?

I was experimenting with this and think its worth following up on this idea. However there are some fields that we set based on the finality notifications (even non-mandatory) and since I am just getting familiar with the code, I opted for the simplest-yet-working approach.

Reply:
Yes we can follow-up with micro-optimizations like removing the extra async task in the middle once initialization is complete. For now, this is good for fixing the issue.

PR just needs prdoc and is good to go from my point of view.

Reply:
Will add that as soon as my nodes synced correctly. Then we can merge 👍